File: python/examples/pull_callback_cross.py
#!/usr/bin/env python3
import argparse
import asyncio
import signal
import sys
from collections.abc import AsyncGenerator
from typing import Any

from camera_ui_rpc import create_rpc_client, rpc_callbacks
from camera_ui_rpc.handler import CallbackInvoker


async def pull_batches(
    batch_count: int, chunks_per_batch: int, invoker: CallbackInvoker
) -> AsyncGenerator[None, None]:
    for b in range(batch_count):
        if not invoker.active:
            return
        for i in range(chunks_per_batch):
            await invoker.invoke("onChunk", {"batch": b, "index": i})
        await invoker.invoke("onChunk", None)
        yield


async def run_server(name: str) -> None:
    print(f"[py-server {name}] starting...")
    client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": f"pullcb-cross-py-server-{name}",
        }
    )
    await client.connect()
    # Register under camelCase subject so Node/Go clients can reach it.
    handlers = {"pullBatches": pull_batches}
    unsub = await client.register_handler(f"pullcb-{name}", handlers, without_decorators=True)
    print(f"[py-server {name}] registered under namespace pullcb-{name}, ready.")

    stop_event = asyncio.Event()

    def _stop() -> None:
        stop_event.set()

    loop = asyncio.get_event_loop()
    for sig in (signal.SIGTERM, signal.SIGINT):
        loop.add_signal_handler(sig, _stop)

    await stop_event.wait()
    print(f"[py-server {name}] shutting down...")
    await unsub()
    await client.disconnect()


async def test_target(client_name: str, target: str) -> bool:
    client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": f"pullcb-cross-py-client-{client_name}-to-{target}",
        }
    )
    try:
        await client.connect()
        proxy = client.create_proxy(f"pullcb-{target}")

        received: list[dict[str, Any]] = []
        batch_ends = 0

        def on_chunk(data: Any) -> None:
            nonlocal batch_ends
            if data is None:
                batch_ends += 1
                return
            received.append(data)

        cbs = rpc_callbacks(onChunk=on_chunk, oneway=["onChunk"])

        BATCHES = 3
        CHUNKS = 5

        batches_consumed = 0
        # camelCase method name — matches Node/Go conventions and the
        # camelCase key in the server's handler dict.
        async for _ in proxy.pullBatches(BATCHES, CHUNKS, cbs):
            batches_consumed += 1

        await asyncio.sleep(0.1)

        order_ok = all(
            int(r.get("batch", -1)) == i // CHUNKS and int(r.get("index", -1)) == i % CHUNKS
            for i, r in enumerate(received)
        )

        ok = (
            batches_consumed == BATCHES
            and len(received) == BATCHES * CHUNKS
            and batch_ends == BATCHES
            and order_ok
        )

        if ok:
            print(f"  [py -> {target}] PASS ({len(received)} chunks, {batch_ends} EOB)")
        else:
            print(
                f"  [py -> {target}] FAIL: consumed={batches_consumed}/{BATCHES}, "
                f"chunks={len(received)}/{BATCHES * CHUNKS}, ends={batch_ends}/{BATCHES}, order_ok={order_ok}"
            )
        return ok
    except Exception as e:
        print(f"  [py -> {target}] ERROR: {e}")
        return False
    finally:
        await client.disconnect()


async def run_client(targets: list[str]) -> None:
    print(f"[py-client] testing targets: {', '.join(targets)}")
    results: list[tuple[str, bool]] = []
    for t in targets:
        ok = await test_target("py", t)
        results.append((t, ok))

    failed = [t for t, ok in results if not ok]
    print()
    if not failed:
        print(f"[py-client] all {len(results)} targets passed")
        sys.exit(0)
    else:
        print(f"[py-client] {len(failed)}/{len(results)} targets failed")
        sys.exit(1)


async def _main(argv: list[str]) -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--role", choices=["server", "client"], default="client")
    parser.add_argument("--name", default="python")
    parser.add_argument("--targets", default="node,go,python")
    args = parser.parse_args(argv)

    if args.role == "server":
        await run_server(args.name)
    else:
        targets = [t.strip() for t in args.targets.split(",") if t.strip()]
        await run_client(targets)


async def main(argv: list[str] | None = None) -> None:
    await _main(argv if argv is not None else sys.argv[1:])


if __name__ == "__main__":
    asyncio.run(main())

File: node/examples/pull-callback-cross.ts
import { createRPCClient, rpcCallbacks } from '../src/index.js';

interface DataService {
  pullBatches(batchCount: number, chunksPerBatch: number, callbacks: { onChunk: (data: { batch: number; index: number } | null) => void }): AsyncGenerator<void>;
}

class DataServiceImpl implements DataService {
  async *pullBatches(batchCount: number, chunksPerBatch: number, callbacks: { onChunk: (data: { batch: number; index: number } | null) => void }) {
    for (let b = 0; b < batchCount; b++) {
      for (let i = 0; i < chunksPerBatch; i++) {
        callbacks.onChunk({ batch: b, index: i });
      }
      callbacks.onChunk(null);
      yield;
    }
  }
}

function parseFlag(argv: string[], name: string): string | undefined {
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === name && i + 1 < argv.length) return argv[i + 1];
    if (argv[i].startsWith(`${name}=`)) return argv[i].slice(name.length + 1);
  }
  return undefined;
}

async function runServer(name: string) {
  console.log(`server ${name} starting...`);
  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: `pullcb-cross-node-server-${name}`,
  });
  await client.connect();
  const unsub = await client.registerHandler(`pullcb-${name}`, new DataServiceImpl(), { withoutDecorators: true });
  console.log(`server ${name} registered under namespace pullcb-${name}, ready.`);

  await new Promise<void>((resolve) => {
    const stop = async () => {
      console.log(`server ${name} shutting down...`);
      await unsub();
      await client.disconnect();
      resolve();
    };
    process.once('SIGTERM', stop);
    process.once('SIGINT', stop);
  });
}

async function testTarget(clientName: string, target: string): Promise<boolean> {
  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: `pullcb-cross-node-client-${clientName}-to-${target}`,
  });
  await client.connect();

  try {
    const service = client.createProxy<DataService>(`pullcb-${target}`);

    const received: { batch: number; index: number }[] = [];
    let batchEnds = 0;

    const cbs = rpcCallbacks(
      {
        onChunk: (data: { batch: number; index: number } | null) => {
          if (data === null) {
            batchEnds++;
            return;
          }
          received.push(data);
        },
      },
      { oneway: ['onChunk'] },
    );

    const BATCHES = 3;
    const CHUNKS = 5;

    let batchesConsumed = 0;
    for await (const _ of service.pullBatches(BATCHES, CHUNKS, cbs)) {
      batchesConsumed++;
    }

    // Allow trailing callback messages to land.
    await new Promise((r) => setTimeout(r, 100));

    const orderOk = received.every((r, i) => {
      const expectedBatch = Math.floor(i / CHUNKS);
      const expectedIdx = i % CHUNKS;
      return r.batch === expectedBatch && r.index === expectedIdx;
    });

    const ok = batchesConsumed === BATCHES && received.length === BATCHES * CHUNKS && batchEnds === BATCHES && orderOk;

    if (ok) {
      console.log(`  node -> ${target}: PASS (${received.length} chunks, ${batchEnds} EOB)`);
    } else {
      console.log(
        // eslint-disable-next-line @stylistic/max-len
        `  node -> ${target}: FAIL: consumed=${batchesConsumed}/${BATCHES}, chunks=${received.length}/${BATCHES * CHUNKS}, ends=${batchEnds}/${BATCHES}, orderOk=${orderOk}`,
      );
    }
    return ok;
  } finally {
    await client.disconnect();
  }
}

async function runClient(targets: string[]) {
  console.log(`testing targets: ${targets.join(', ')}`);
  const results: { target: string; ok: boolean }[] = [];
  for (const t of targets) {
    try {
      const ok = await testTarget('node', t);
      results.push({ target: t, ok });
    } catch (err) {
      console.log(`  node -> ${t}: ERROR: ${(err as Error).message}`);
      results.push({ target: t, ok: false });
    }
  }

  const failed = results.filter((r) => !r.ok);
  console.log('');
  if (failed.length === 0) {
    console.log(`all ${results.length} targets passed`);
    process.exit(0);
  } else {
    console.log(`${failed.length}/${results.length} targets failed`);
    process.exit(1);
  }
}

async function main() {
  const role = parseFlag(process.argv.slice(2), '--role') ?? 'client';
  if (role === 'server') {
    const name = parseFlag(process.argv.slice(2), '--name') ?? 'node';
    await runServer(name);
  } else {
    const targetsArg = parseFlag(process.argv.slice(2), '--targets') ?? 'node,go,python';
    const targets = targetsArg
      .split(',')
      .map((t) => t.trim())
      .filter(Boolean);
    await runClient(targets);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(2);
});
